### PR TITLE
Fix superfluous TOC. dependency on ppc64 (bsc#1113100), Add kernel module provides kmod(modulename) (FATE#326579). Add automatic kernel module requires for module-load.d files

### PR DESCRIPTION
**Pull request by hramrach**
This resurrects code from earlier version of the script that eliminates
depends that are also provided by the package. Since the TOC. is
provided by everything this removes it.

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
@hramrach Please add a modern style changes entry. :)

**Comment by hramrach** [comment]
please specify what is modern style changes entry

**Comment by Conan-Kudo** [comment]
A changes entry in `package/rpm-config-SUSE.changes` formatted like so:

```
-------------------------------------------------------------------
Mon Feb 26 03:48:20 UTC 2018 - Neal Gompa <ngompa13@gmail.com>

- Initial split of RPM vendor configuration from rpm package

```

**Comment by hramrach** [comment]
That's doable in general. Question is what all needs to be parsed. And packages that load modules in a package-specific way will need explicit dependency as well. Another thing is that some packages might use modules.load.d to load some modules in case they are available but not as a hard requirement.

**Comment by mwilck** [comment]
> That's doable in general. Question is what all needs to be parsed. And packages that load modules in a package-specific way will need explicit dependency as well. 

Simply catching modules-load.d style dependencies would be a good start. IMO fully sufficient for SLE15-SP1.

> Another thing is that some packages might use modules.load.d to load some modules in case they are available but not as a hard requirement.

IMO using modules-load.d in that case would be wrong. It causes error messages on every boot. See [bsc1119414](https://bugzilla.suse.com/show_bug.cgi?id=1119414).

**Comment by mwilck** [comment]
We can't simply add this to Factory.

If no code for auto-generated "Requires:" is introduced, and this code is added to Factory, **%suse_version** must be increased, or some other RPM macro must be defined that allows to determine whether this feature is enabled. Packagers need to be able to write code in spec files that determines if kmod(xyz)-style provides are available. Otherwise we'll get broken dependencies all over the place.

On SLE, we can use **%sle_version**, but this needs to work on OpenSUSE, in particular Tumbleweed, too.

**Comment by hramrach** [comment]
Once it's added to Factory you can assume it's in Factory. And you can use %suse_version elsewhere.

The situation with modules.load.d is a bit tricky.

If a package installs a module and adds it to modules.load.d we must not add it. Otherwise we get the same situation as the .TOC

**Comment by mwilck** [comment]
> Once it's added to Factory you can assume it's in Factory. And you can use %suse_version elsewhere.

What about staging projects? Factory snapshots? I still think the presence of this feature must be detectable at build time. What would be the reason **not** to increase %suse_version in factory?

> The situation with modules.load.d is a bit tricky.
> 
> If a package installs a module and adds it to modules.load.d we must not add it. Otherwise we get the same situation as the .TOC

Hm, isn't rpm able to handle stuff like this?

    Requires: foo-1.0
    Provides: foo-1.0

**Comment by DimStar77** [comment]
> What about staging projects? Factory snapshots? I still think the presence of this feature must be detectable at build time. What would be the reason not to increase %suse_version in factory?

Not sure why you would need to detect this: you produce provides if the generator is present, and you don't produce any if this code is not yet there. If at a later stage, you bring code to add requires, Factory clearly moved far enough already - and if you're faster than that, we just add both pieces into one staging (all others have no info about it)

Bumping suse_version is not really the way out - we are defined at 1550 in Tumbleweed, - between SLE 15 (1500) and the upcoming SLE16 (1600). There is more to it than 'just' changing the version in the prjconf (which would even be bumped for Factory and Tumbleweed at the same time - as this is the same project in OBS)

Stagings are an implementation detail and are generally the same as Factory - and Tumbleweed is nothing more than a Factory snapshot that we tested and considered 'good enough to publish'

**Comment by mwilck** [comment]
> Not sure why you would need to detect this: you produce provides if the generator is present, and you don't produce any if this code is not yet there.

It's about detecting whether I should put 

     Requires: kmod(xyz)

in a package. If I do that, the package won't install on any system where the kernel had been generated with and older "find-provides" version. So I need to place these requires in an %if conditional. How am I supposed to write that conditional? I can check for SLE15-SP1 with %sle-version, but what check do I write for factory? IOW, if I generate these `Requires:` for `0%{suse_version} >= 1550` now, the package won't install any more on Factory. I wouldn't want that for my current submissions. I'd rather wait until I know the "provides" generator is in place, but how would I know without polling? I have to admit I don't read every new Tumbleweed announcement.

> Bumping suse_version is not really the way out - we are defined at 1550 in Tumbleweed, - between SLE 15 (1500) and the upcoming SLE16 (1600). 

Sorry for being naïve. I thought the trailing zeroes were there for some purpose, and we could bump %suse_version e.g. to 1510. But if that's impossible, perhaps we can define e.g. `%{__kmod_provides__}` in the project instead?

**Comment by hramrach** [comment]
Defining something in project is not going to work. It needs defined on every new release, PTF, home project, devel project, and whatnot. If your package submission requires the new provides it should be lumped together with rpm-config-suse in a staging project. You may not follow Tumbleweed development but Tumbleweed maintainers should.

> 
> Hm, isn't rpm able to handle stuff like this?
> 
```
> Requires: foo-1.0
> Provides: foo-1.0
```
Apparently it cannot. Otherwise we would not have the .TOC bug. It can be filtered in the script.

The other question is how do we deliver this to SLE/Leap. There is one rpm package per codestream and no separate rpm-config-SUSE. So if we add these automatic requires we will need to rebuild all existing KMPs. (ie if we want automagic requires in 15 SP1 we need to rebuild 15 GA and Update KMPs).

**Comment by DimStar77** [comment]
you overcomplicate live - if you submit today a package that has a Requires: kmod(), the bot will refuse it from entering until we have the code in the repo that does the respective Provides: part; There is literally no reason to overcomplicate anybody's life here. Simply submit the code to add the provides - and in the packages having the requires guard them with suse_version >= 1550 (meaning Tumbleweed plus future SLE versions)

**Comment by hramrach** [comment]
ok, so do we want the automagic kmod() requires, and where?

I don't see any problem with them in Tumbleweed. With SLE/Leap the situation is more tricky. We do have the kmod() provides for a while so recent kernel MUs and KMP updates should have them. Is it acceptable to just add the requires in newly built packages and then lazily rebuild any KMPs when a kmod() require is not satisfied because the KMP was not updated recently?

**Comment by mwilck** [comment]
> Simply submit the code to add the provides - and in the packages having the requires guard them with suse_version >= 1550

Nope. I need to make submissions for multipath-tools now. I can't wait for this "provides" code to be merged. "Factory first" forces me to have the packages submitted Factory before they go to SLE15-SP1. Factory submissions declined by a bot don't seem fit as a base for a SLE submission.

For me the result of this discussion is to guard the the "Requires:" with `0%{? sle_version >= 150100}` for the time being. Sorry for overcomplicating life.

**Comment by hramrach** [comment]
Or just fork multipath-tools in SLE and blame it on people not merging this fast enough. It is rotting here for *months*. Factory first assumes responsive maintainers.

**Comment by mwilck** [comment]
> ok, so do we want the automagic kmod() requires, and where?

I'm for it, but it turns out that I don't need it, for multipath-tools at least. I created a macro to construct the dependencies from the `modules-load.d` file I ship. See [the spec file, line 90ff.](https://build.opensuse.org/package/view_file/home:mwilck:branches:Base:System/multipath-tools/multipath-tools.spec?expand=1).

**Comment by DimStar77** [comment]
> For me the result of this discussion is to guard the the "Requires:" with 0%{? sle_version >= 150100} for the time being. Sorry for overcomplicating life.

Just in case this not all readers are aware: Tumbleweed has no sle_version defined (so above statement would exclude Tumbleweed) - and if SLE was to follow Factory-First, you could not get that into SLE15SP1 much sooner (at least submissions would need to exist to both products)

**Comment by mwilck** [comment]
FTR, https://build.opensuse.org/request/show/681792. The SLE15-SP1 package is going to be forked anyway, but I strongly prefer to keep the spec files in sync.

**Comment by hramrach** [comment]
Here it is. Should be fine for Tumbleweed.

**Comment by mwilck** [comment]
A word of caution:

I have to report a trap I fell into with multipath-tools. Parsing `modules-load.d` files may produce wrong dependencies if the entries in the files don't match file names exactly. For example, we had this in the multipath.tools package in a `modules-load.d` file:

    scsi_dh_alua
    scsi_dh_emc
    scsi_dh_rdac
    dm_multipath

Converting this innocent-looking and correctly working code into `kmod()` dependencies fails because the requirement `kmod(dm_multipath)` can't be satisfied. The reason is that the file name of the module is `dm-multipath.ko`, not `dm_multipath.ko`. **modprobe** has no problem with this, as it treats dashes and underscores the same character. But the same is of course not true for rpm.

This can of course be avoided by using the proper module file names in `modules-load.d` files. But because **modprobe** deals gracefully with typos like this, it's not unlikely that other packages contain similar mistakes.

**Comment by hramrach** [comment]
> Shouldn't you skip empty lines in modules.d files, too?
> 
> ```
>   case "$x" in \#*|\;*|"");;
> ```
> What about `/etc/modules-load.d` ?

It should be handled by the grep matching empty lines.

> A word of caution:
> 
> I have to report a trap I fell into with multipath-tools. Parsing `modules-load.d` files may produce wrong dependencies if the entries in the files don't match file names exactly. For example, we had this in the multipath.tools package in a `modules-load.d` file:
> 
> ```
> scsi_dh_alua
> scsi_dh_emc
> scsi_dh_rdac
> dm_multipath
> ```
> Converting this innocent-looking and correctly working code into `kmod()` dependencies fails because the requirement `kmod(dm_multipath)` can't be satisfied. The reason is that the file name of the module is `dm-multipath.ko`, not `dm_multipath.ko`. **modprobe** has no problem with this, as it treats dashes and underscores the same character. But the same is of course not true for rpm.

Maybe we can convert dashes to underscores in both provides and requires.

**Comment by mwilck** [comment]
> Maybe we can convert dashes to underscores in both provides and requires.

Looks good to me, except that you still don't handle `/etc/modules-load.d`. But maybe that's not so bad after all. It would offer an opportunity to avoid modules-load.d files being auto-converted into kmod-dependencies.

**Comment by mwilck** [comment]
The only problem about the dash/underscore conversion is that it's currently handled differently in SLE15-SP1. Changing that would result in dependency nightmares. But having different provides / requires in Factory and SLE15 is also sub-optimal.

**Comment by hramrach** [comment]
I submitted update to change to dashes in SLE15 as well. So long as it gets there before GA we can re-adjust the dependencies.

/etc/ is reserved for administrator so packages installing files in /etc/modules-load.d should handle their dependencies. It is trivial to add handling /etc/ as well if there is a good reason.

**Comment by mlschroe** [comment]
Is everybody ok with this now? Can it go into Factory?

**Comment by mwilck** [comment]
It's OK for me. 

Do we have a list of packages shipping modules-load.d files? We may want to inform the maintainers that their packages are going to get additional Requires: tags, which might cause certain image builds to blow up (bug 1119414). It might catch some people by surprise otherwise.

**Comment by gmoro** [comment]
> It's OK for me.
> 
> Do we have a list of packages shipping modules-load.d files? We may want to inform the maintainers that their packages are going to get additional Requires: tags, which might cause certain image builds to blow up (bug 1119414). It might catch some people by surprise otherwise.
Sorry for the verbosity, I didn't really cleaned it up, but here it is a list I got from SLE15-SP1 ISO some weeks ago.

`/usr/lib/modules-load.d
/usr/lib/modules-load.d/ecryptfs.conf
./install/x86_64/ecryptfs-utils-111-2.31.x86_64.rpm
ecryptfs-utils-111-2.31.x86_64
---
/usr/lib/modules-load.d
/usr/lib/modules-load.d/multipath.conf
./install/x86_64/multipath-tools-0.7.9+111+suse.b4232b7-4.9.x86_64.rpm
multipath-tools-0.7.9+111+suse.b4232b7-4.9.x86_64
---
/etc/modules-load.d
/etc/modules-load.d/sg.conf
./install/x86_64/suse-module-tools-15.1.10-1.1.x86_64.rpm
suse-module-tools-15.1.10-1.1.x86_64
---
/etc/modules-load.d
/usr/lib/modules-load.d
/usr/lib/modules-load.d/sg.conf
/usr/lib/systemd/system/sysinit.target.wants/systemd-modules-load.service
/usr/lib/systemd/system/systemd-modules-load.service
/usr/lib/systemd/systemd-modules-load
/usr/share/man/man5/modules-load.d.5.gz
/usr/share/man/man8/systemd-modules-load.8.gz
/usr/share/man/man8/systemd-modules-load.service.8.gz
./install/x86_64/systemd-234-24.20.1.x86_64.rpm
systemd-234-24.20.1.x86_64
---
/usr/lib/modules-load.d
/usr/lib/modules-load.d/ecryptfs.conf
./packages/SP1-Module-Basesystem/x86_64/ecryptfs-utils-111-2.31.x86_64.rpm
ecryptfs-utils-111-2.31.x86_64
---
/usr/lib/modules-load.d
/usr/lib/modules-load.d/multipath.conf
./packages/SP1-Module-Basesystem/x86_64/multipath-tools-0.7.9+111+suse.b4232b7-4.9.x86_64.rpm
multipath-tools-0.7.9+111+suse.b4232b7-4.9.x86_64
---
/etc/modules-load.d
/etc/modules-load.d/sg.conf
./packages/SP1-Module-Basesystem/x86_64/suse-module-tools-15.1.10-1.1.x86_64.rpm
suse-module-tools-15.1.10-1.1.x86_64
---
/etc/modules-load.d
/usr/lib/modules-load.d
/usr/lib/modules-load.d/sg.conf
/usr/lib/systemd/system/sysinit.target.wants/systemd-modules-load.service
/usr/lib/systemd/system/systemd-modules-load.service
/usr/lib/systemd/systemd-modules-load
/usr/share/man/man5/modules-load.d.5.gz
/usr/share/man/man8/systemd-modules-load.8.gz
/usr/share/man/man8/systemd-modules-load.service.8.gz
./packages/SP1-Module-Basesystem/x86_64/systemd-234-24.20.1.x86_64.rpm
systemd-234-24.20.1.x86_64
---
/usr/lib/modules-load.d
./packages/SP1-Module-Server-Applications/x86_64/xen-tools-4.12.0_02-1.10.x86_64.rpm
xen-tools-4.12.0_02-1.10.x86_64
---`

**Comment by mwilck** [comment]
Note that we still need consensus on how the kmod() dependencies should be implemented in detail, in particular if they should be conditional / use extended syntax ([SLE-3853](https://jira.suse.de/browse/SLE-3853)).

File: scripts/find-requires.ksyms [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 IFS=$'\n'
 
+case "$1" in
+kernel-module-*)    ;; # Fedora kernel module package names start with
+                      # kernel-module.
```
Comment (Conan-Kudo) [inline scripts/find-requires.ksyms]: RH/Fedora kernel module packages start with `kmod-`... They use virtual provides `kmod(<kmod>)`.

Cf. https://rpmfusion.org/Packaging/KernelModules/Kmods2

**Comment by Conan-Kudo** [comment]
I'm going to pretend you didn't just link to a resource people can't access (including myself) and try to answer the question anyway.

I don't think there's a good reason for implementing kmod requires right now, but if we _want_ it, then it should probably be simple `Requires: kmod(<kmod>)` from files in `%_modulesloaddir`, preserving the case-sensitivity listed there.

That means, yes, there may be broken dependencies, but at least you'll _know_ and you can fix it.

**Comment by Conan-Kudo** [comment]
Also, since I notice it being slightly referenced in this PR, for future reference, RH/Fedora kernel module packages are generated using [`kmodtool`](https://src.fedoraproject.org/rpms/kmodtool).

There's [another variant for RHEL included in redhat-rpm-config](https://src.fedoraproject.org/rpms/redhat-rpm-config/blob/master/f/kmodtool).

Both variants can be used with [the kernel module package macros](https://src.fedoraproject.org/rpms/redhat-rpm-config/blob/master/f/macros.kmp) that work similarly to the SUSE ones.

**Comment by mwilck** [comment]
> I'm going to pretend you didn't just link to a resource people can't access (including myself)

Sorry if this is causing frustration on your part. It was meant as a reminder for those people who'd been involved in this discussion until you joined.

> [...] but if we want it, then it should probably be simple Requires: kmod(<kmod>) from files in %_modulesloaddir, preserving the case-sensitivity listed there.
> That means, yes, there may be broken dependencies, but at least you'll know and you can fix it.

I agree. Other people see it differently.

> RH/Fedora kernel module packages start with `kmod-...`

Good point. Maybe we should come up with a different naming convention for these new dependencies, which are not the same thing as RH's `kmod`.

**Comment by Conan-Kudo** [comment]
> Sorry if this is causing frustration on your part. It was meant as a reminder for those people who'd been involved in this discussion until you joined.

It might not have been obvious (due to things that I don't really want to get into), but I created this project and am the main maintainer. I've mostly stayed out of this discussion because there wasn't enough information for me to give any feedback on.

> Good point. Maybe we should come up with a different naming convention for these new dependencies, which are not the same thing as RH's kmod.

What makes these different? It seems like they are, in fact, the same. I'd really like for there to be more convergence between RH/Fedora and (open)SUSE on kernel module packaging because it would make the whole thing a lot less confusing. It's not like you're solving different problems, after all.

**Comment by mwilck** [comment]
> > Good point. Maybe we should come up with a different naming convention for these new dependencies, which are not the same thing as RH's kmod.
> 
> What makes these different? It seems like they are, in fact, the same. I'd really like for there to be more convergence between RH/Fedora and (open)SUSE on kernel module packaging because it would make the whole thing a lot less confusing. It's not like you're solving different problems, after all.

The semantics are very different. Fedora / RH uses `kmod` for **kernel module packages**, as you point out correctly. SUSE wants to use `kmod` for **built-in kernel modules**.

On RHEL, a kernel module package provides (epoch omitted for readability):

    %{name}-kmod = %{version}-%{release}
    kmod-%{name}%{?flavor:-%{flavor}} = %{version}-%{release} 
    kmod-%{name}%{?flavor:-%{flavor}}(%{arch}) = %{version}-%{release}

This corresponds to these provides of SUSE kernel module packages:

    %{name}-kmp = %{version}-%{release}
    %{name}-kmp-%{flavor} = %{version}-%{release}
    %{name}-kmp.%{flavor}(%{arch}) = %{version}-%{release}

With the changes discussed in this PR, a SUSE kernel module package also provides

     kmod(%{name})

(without version - personally I'd actually appreciate if we'd make this a versioned `Provides:`, but that's a different discussion).

The SUSE kernel package provides `%{name}-kmp` and `%{name}-kmp-%{flavor}` only for a few drivers that have previously been shipped as KMPs, but it will have `kmod()` provides for **every module it contains**.

Arguably, there's some overlap between our existing `kmp-` provides and the new `kmod()` ones. But the purpose is very different. But anyway, SUSE using `kmod` for a different thing than RH may cause confusion among users; that's what I wanted to say.

**Comment by Conan-Kudo** [comment]
> The semantics are very different. Fedora / RH uses kmod for kernel module packages, as you point out correctly. SUSE wants to use kmod for built-in kernel modules.

RH/Fedora uses it for both. kernel packages provide kmod() provides:
* https://koji.fedoraproject.org/koji/rpminfo?rpmID=17361378
* https://koji.fedoraproject.org/koji/rpminfo?rpmID=17361382

Although, now I understand what you mean by `kernel-modules-` now. Yes, the kernel package does produce `kernel-modules-` packages.

**Comment by mwilck** [comment]
> RH/Fedora uses it for both. kernel packages provide kmod() provides:

OK, I didn't know that. Then, our use of `kmod` wouldn't be so different (except that we omit the `.ko`).

**Comment by Conan-Kudo** [comment]
@mwilck Well, would you be open to not omitting the `.ko`? It'd be nice to uniform kmod() provides?

**Comment by mwilck** [comment]
> @mwilck Well, would you be open to not omitting the `.ko`?

That's rather a question for @hramrach. I don't have strong feelings about it.

**Comment by hramrach** [comment]
And will we provide .ko.gz when the modules are compressed?
Either way, we are in GMC with SP1 already providing kmod() without .ko
Feel free to patch it to provide with .ko as well.
Dropping .ko now will probably not be welcome.

**Comment by Conan-Kudo** [comment]
No, compressed extensions would be dropped. Fedora does that now.

**Comment by hramrach** [comment]
I hope this does it. With the package split different between different codestreams it is very difficult to verify :/

**Comment by Conan-Kudo** [comment]
@hramrach @mwilck @DimStar77 If we're all happy with this, I can merge it?

**Comment by mwilck** [comment]
I hate to be the notorious sceptic again, but is it really prudent to provide both `kmod(x)` and `kmod(x.ko)`? Shouldn't we just decide on one? We'd have ~4000 `kmod`-provides in kernel-default anyway, and this would double the number...

**Comment by Conan-Kudo** [comment]
@mwilck We technically don't need to, but I think @hramrach wants to for compatibility with legacy SLE.

**Comment by hramrach** [comment]
We will need to branch rpm to introduce the automatic dependencies anyway so there is not much problem for new SLE.

The packages that added manually dependencies on kmod(x) without .ko in 15 SP1 are another problem. They will break once the automatic dependency is introduced but I suppose there will be enough time to fix them once that happens.

**Comment by Conan-Kudo** [comment]
@hramrach So do you want to drop the `kmod(foo)` provides then?

**Comment by Vogtinator** [comment]
If I'm not mistaken this will cause the same issue which `multipath-tools` ran into. `installation-images` builds an initrd and a bit more to start the installation system, including `multipath-tools` for multipathing support during installation. The result was that the installation system included a full copy of `kernel-default` due to the kmod requirement, resulting in broken installations.

This is unfortunately being discussed in a private bug report...

**Comment by mwilck** [comment]
@Vogtinator, this is not going to happen, at least not with multipath-tools, as TW already has the package where libmpath0 has been split out. So small images which don't need the actual multipath functionality should simply not pull in multipath-tools any more.

**Comment by Vogtinator** [comment]
@mwilck: The split does not make any difference with installation-images as it pulls in `multipath-tools` as it actually needs multipathing support.

**Comment by Vogtinator** [comment]
@mwilck: Yes, but kernel-default is provided by the "outside", not installation-images itself. For details ask @wfeldt.

**Comment by mwilck** [comment]
Fine, but that's really a very special requirement of the installation images, and I'm sure a workaround could be found very easily. It could simply be a pseudo-package containing the missing `Provides:`.

**Pull request merged**